### ISSUE TITLE
`struct termios` is broken on Linux

### ISSUE THREAD

**Issue by mbr** [dcuddeback/termios-rs#9]
There seems to be a bit of a misunderstanding of the central structure of the serial library either by me or the library; I've been chasing down this particular bug for *hours* now (this issue is also relevant for #8).

The wrong datastructure is labelled as `pub struct termios`, it really should be `struct termios2`. The real issue is a different one though: The `NCSS` constant is wrong!

Compare the [current source](https://github.com/dcuddeback/termios-rs/blob/696b03c6f2412e320eae436aa76ccd8011b4d3a6/src/os/linux.rs#L24)

```rust
pub const NCCS: usize = 32;
```

```rust
#[derive(Debug,Copy,Clone,Eq,PartialEq)]
#[repr(C)]
pub struct termios {
    pub c_iflag: tcflag_t,
    pub c_oflag: tcflag_t,
    pub c_cflag: tcflag_t,
    pub c_lflag: tcflag_t,
    c_line: cc_t,
    pub c_cc: [cc_t; NCCS],
    c_ispeed: speed_t,
    c_ospeed: speed_t
}
```

resulting in a struct with size (4*4 + 1 [+ 3] + 32 * 1 + 4 + 4 =) 60 bytes or **0x3c** with the [C headers](http://lxr.free-electrons.com/source/include/uapi/asm-generic/termbits.h#L10):

```c
#define NCCS 19
```

```c
struct termios2 {
        tcflag_t c_iflag;               /* input mode flags */
        tcflag_t c_oflag;               /* output mode flags */
        tcflag_t c_cflag;               /* control mode flags */
        tcflag_t c_lflag;               /* local mode flags */
        cc_t c_line;                    /* line discipline */
        cc_t c_cc[NCCS];                /* control characters */
        speed_t c_ispeed;               /* input speed */
        speed_t c_ospeed;               /* output speed */
};
```

The resulting struct here is only (4*4 + 1 [+ 0] + 19 * 1 + 4 + 4) = 44 or **0x2c** bytes long. Compare with the correct syscall to fill it: `TCGETS2` == `0x802c542a`; which indicates a read call of size `0x002c`.

In short:

* `termios-rs` uses the wrong datastructure or ioctls (depending on what one wants to support) for its calls.
* The datastructure is also too large as well, causing a the extended speed fields to be "pushed" outside the region that is written by kernel syscalls.
* Because only the first part of each structure is used, this goes unnoticed, until extended features of this structure are used.

**Comment by dcuddeback** [comment]
@mbr You're looking at the wrong header files. `struct termios` is defined in glibc: http://sourceware.org/git/?p=glibc.git;a=blob;f=bits/termios.h;h=5487b514c67d27ebc30a41e3cabdfddfad971d9d;hb=refs/heads/master#l110

Compiling the following C program agrees with the definitions in this crate:

```c
#include <stdio.h>
#include <termios.h>

int main(void)
{
    printf("NCCS = %d\n", NCCS);
    printf("sizeof(termios) = %lu\n", sizeof(struct termios));
    return 0;
}
```

Output:

```
NCCS = 32
sizeof(termios) = 60
```

**Comment by mbr** [comment]
> @mbr You're looking at the wrong header files. struct termios is defined in glibc:

My god, that glibc `struct termios` is an unreadable mess.

That aside, here's my take on the example program:

```c
#include <stdio.h>
#include <asm/termios.h>


int main(void)
{
    printf("NCCS = %d\n", NCCS);
    printf("sizeof(termios2) = %lu\n", sizeof(struct termios2));
    return 0;
}
```

yields

```
marc@marc-desktop /tmp> gcc sztest.c 
marc@marc-desktop /tmp> ./a.out 
NCCS = 19
sizeof(termios2) = 44
```

So using a different header produces different results. My gripe here is that, while the manpage uses `termios.h`, using `asm/termios.h` does work - while the former does not.

I did test both struct sizes and measures the resulting custom baudrate with an actual oscilloscope; only NCCS = 19 worked.

I wonder why I did not link to the Stack Overflow issue here that describes this phenomena, here's the money quote from http://stackoverflow.com/questions/12646324/how-to-set-a-custom-baud-rate-on-linux

> I noticed the same thing about BOTHER not being defined. Like Jamey Sharp said, you can find it in <asm/termios.h>. Just a forewarning, I think I ran into problems including both it and the regular <termios.h> file at the same time.

> Aside from that, I found with the glibc I have, it still didn't work because glibc's tcsetattr was doing the ioctl for the old-style version of struct termios which doesn't pay attention to the speed setting. I was able to set a custom speed by manually doing an ioctl with the new style termios2 struct, which should also be available by including <asm/termios.h>:

In summary, it seems that glibc is broken and as a result, termios-rs is broken. You could try the example code on the StackOverflow answer; it does work for me (verifying with stty shows a baudrate if `0`, but it is correct when sending).
